File: jane_street/electronic_trading_challenge/cpp/src/strategies.cpp
#include "etc/strategies.h"
#include <numeric>
#include <algorithm>

namespace etc {

std::vector<Action> BondStrategy::execute(const MarketState& state) {
    std::vector<Action> actions;
    auto it = state.books.find("BOND");
    if (it == state.books.end()) return actions;

    int current_pos = 0;
    auto pos_it = state.positions.find("BOND");
    if (pos_it != state.positions.end()) current_pos = pos_it->second;

    int limit = 100;
    if (current_pos < limit) {
        actions.push_back({"add", "BOND", "BUY", 999, limit - current_pos});
    }
    if (current_pos > -limit) {
        actions.push_back({"add", "BOND", "SELL", 1001, limit + current_pos});
    }
    return actions;
}

std::vector<Action> AdrStrategy::execute(const MarketState& state) {
    std::vector<Action> actions;
    auto valbz_it = state.last_trades.find("VALBZ");
    auto vale_it = state.last_trades.find("VALE");

    if (valbz_it == state.last_trades.end() || vale_it == state.last_trades.end()) return actions;
    if (valbz_it->second.size() < 10 || vale_it->second.size() < 10) return actions;

    double valbz_mean = std::accumulate(valbz_it->second.end() - 10, valbz_it->second.end(), 0.0) / 10.0;
    double vale_mean = std::accumulate(vale_it->second.end() - 10, vale_it->second.end(), 0.0) / 10.0;

    if (valbz_mean - vale_mean >= 2) {
        actions.push_back({"add", "VALE", "BUY", (int)vale_mean + 1, 10});
        actions.push_back({"convert", "VALE", "SELL", 0, 10});
        actions.push_back({"add", "VALBZ", "SELL", (int)valbz_mean - 1, 10});
    }
    return actions;
}

std::vector<Action> EtfStrategy::execute(const MarketState& state) {
    std::vector<Action> actions;
    std::vector<std::string> symbols = {"XLF", "BOND", "GS", "MS", "WFC"};
    std::map<std::string, double> means;

    for (const auto& sym : symbols) {
        auto it = state.last_trades.find(sym);
        if (it == state.last_trades.end() || it->second.size() < 25) return actions;
        means[sym] = std::accumulate(it->second.end() - 25, it->second.end(), 0.0) / 25.0;
    }

    double nav = 3 * means["BOND"] + 2 * means["GS"] + 3 * means["MS"] + 2 * means["WFC"];
    double xlf_price = 10 * means["XLF"];

    if (xlf_price + 150 < nav) {
        actions.push_back({"add", "XLF", "BUY", (int)means["XLF"] + 1, 100});
        actions.push_back({"convert", "XLF", "SELL", 0, 100});
        actions.push_back({"add", "BOND", "SELL", (int)means["BOND"] - 1, 30});
        actions.push_back({"add", "GS", "SELL", (int)means["GS"] - 1, 20});
        actions.push_back({"add", "MS", "SELL", (int)means["MS"] - 1, 30});
        actions.push_back({"add", "WFC", "SELL", (int)means["WFC"] - 1, 20});
    } else if (xlf_price - 150 > nav) {
        actions.push_back({"add", "BOND", "BUY", (int)means["BOND"] + 1, 30});
        actions.push_back({"add", "GS", "BUY", (int)means["GS"] + 1, 20});
        actions.push_back({"add", "MS", "BUY", (int)means["MS"] + 1, 30});
        actions.push_back({"add", "WFC", "BUY", (int)means["WFC"] + 1, 20});
        actions.push_back({"convert", "XLF", "BUY", 0, 100});
        actions.push_back({"add", "XLF", "SELL", (int)means["XLF"] - 1, 100});
    }

    return actions;
}

} // namespace etc

File: jane_street/electronic_trading_challenge/rust/src/strategies.rs
use crate::models::{MarketState, Action};

pub trait Strategy {
    fn execute(&self, state: &MarketState) -> Vec<Action>;
}

pub struct BondStrategy;
impl Strategy for BondStrategy {
    fn execute(&self, state: &MarketState) -> Vec<Action> {
        let mut actions = vec![];
        if !state.books.contains_key("BOND") { return actions; }
        
        let current_pos = *state.positions.get("BOND").unwrap_or(&0);
        let limit = 100;
        
        if current_pos < limit {
            actions.push(Action::Add { symbol: "BOND".to_string(), dir: "BUY".to_string(), price: 999, size: limit - current_pos });
        }
        if current_pos > -limit {
            actions.push(Action::Add { symbol: "BOND".to_string(), dir: "SELL".to_string(), price: 1001, size: limit + current_pos });
        }
        actions
    }
}

pub struct AdrStrategy;
impl Strategy for AdrStrategy {
    fn execute(&self, state: &MarketState) -> Vec<Action> {
        let mut actions = vec![];
        let valbz_trades = state.last_trades.get("VALBZ");
        let vale_trades = state.last_trades.get("VALE");

        if let (Some(valbz), Some(vale)) = (valbz_trades, vale_trades) {
            if valbz.len() >= 10 && vale.len() >= 10 {
                let valbz_mean: f64 = valbz.iter().rev().take(10).sum::<i32>() as f64 / 10.0;
                let vale_mean: f64 = vale.iter().rev().take(10).sum::<i32>() as f64 / 10.0;

                if valbz_mean - vale_mean >= 2.0 {
                    actions.push(Action::Add { symbol: "VALE".to_string(), dir: "BUY".to_string(), price: vale_mean as i32 + 1, size: 10 });
                    actions.push(Action::Convert { symbol: "VALE".to_string(), dir: "SELL".to_string(), size: 10 });
                    actions.push(Action::Add { symbol: "VALBZ".to_string(), dir: "SELL".to_string(), price: valbz_mean as i32 - 1, size: 10 });
                }
            }
        }
        actions
    }
}

pub struct EtfStrategy;
impl Strategy for EtfStrategy {
    fn execute(&self, state: &MarketState) -> Vec<Action> {
        let mut actions = vec![];
        let symbols = ["XLF", "BOND", "GS", "MS", "WFC"];
        let mut means = std::collections::HashMap::new();

        for &sym in &symbols {
            let trades = match state.last_trades.get(sym) {
                Some(t) if t.len() >= 25 => t,
                _ => return actions,
            };
            let mean: f64 = trades.iter().rev().take(25).sum::<i32>() as f64 / 25.0;
            means.insert(sym, mean);
        }

        let nav = 3.0 * means["BOND"] + 2.0 * means["GS"] + 3.0 * means["MS"] + 2.0 * means["WFC"];
        let xlf_price = 10.0 * means["XLF"];

        if xlf_price + 150.0 < nav {
            actions.push(Action::Add { symbol: "XLF".to_string(), dir: "BUY".to_string(), price: means["XLF"] as i32 + 1, size: 100 });
            actions.push(Action::Convert { symbol: "XLF".to_string(), dir: "SELL".to_string(), size: 100 });
            actions.push(Action::Add { symbol: "BOND".to_string(), dir: "SELL".to_string(), price: means["BOND"] as i32 - 1, size: 30 });
            actions.push(Action::Add { symbol: "GS".to_string(), dir: "SELL".to_string(), price: means["GS"] as i32 - 1, size: 20 });
            actions.push(Action::Add { symbol: "MS".to_string(), dir: "SELL".to_string(), price: means["MS"] as i32 - 1, size: 30 });
            actions.push(Action::Add { symbol: "WFC".to_string(), dir: "SELL".to_string(), price: means["WFC"] as i32 - 1, size: 20 });
        } else if xlf_price - 150.0 > nav {
            actions.push(Action::Add { symbol: "BOND".to_string(), dir: "BUY".to_string(), price: means["BOND"] as i32 + 1, size: 30 });
            actions.push(Action::Add { symbol: "GS".to_string(), dir: "BUY".to_string(), price: means["GS"] as i32 + 1, size: 20 });
            actions.push(Action::Add { symbol: "MS".to_string(), dir: "BUY".to_string(), price: means["MS"] as i32 + 1, size: 30 });
            actions.push(Action::Add { symbol: "WFC".to_string(), dir: "BUY".to_string(), price: means["WFC"] as i32 + 1, size: 20 });
            actions.push(Action::Convert { symbol: "XLF".to_string(), dir: "BUY".to_string(), size: 100 });
            actions.push(Action::Add { symbol: "XLF".to_string(), dir: "SELL".to_string(), price: means["XLF"] as i32 - 1, size: 100 });
        }

        actions
    }
}
